lib/pg_ha_migrations/safe_statements.rb

```diff
@@ -528,40 +536,64 @@ def exec_migration(conn, direction)
+
+    # Lock mode is the same across supplied tables so just grab the first.
+    # This variable is useful for error handling and messaging.
+    target_mode = target_tables.first.mode
 
-    if nested_target_table
-      if nested_target_table != target_table
-        raise PgHaMigrations::InvalidMigrationError, "Nested lock detected! Cannot acquire lock on #{target_table.fully_qualified_name} while #{nested_target_table.fully_qualified_name} is locked."
-      elsif nested_target_table.mode < target_table.mode
-        raise PgHaMigrations::InvalidMigrationError, "Lock escalation detected! Cannot change lock level from :#{nested_target_table.mode} to :#{target_table.mode} for #{target_table.fully_qualified_name}."
+    # Grab the latest locked tables from the call stack.
+    # This will be nil if we are not in a nested context.
+    nested_target_tables = safely_acquire_lock_for_table_history.last
+
+    if nested_target_tables
+      if target_tables != nested_target_tables
+        raise PgHaMigrations::InvalidMigrationError,
+          "Nested lock detected! Cannot acquire lock on #{target_tables.map(&:fully_qualified_name).join(", ")} " \
+          "while #{nested_target_tables.map(&:fully_qualified_name).join(", ")} is locked."
       end
-    else
-      Thread.current[__method__] = target_table
-    end
 
-    # Locking a partitioned table will also lock child tables (including sub-partitions),
-    # so we need to check for blocking queries on those tables as well
-    target_tables = target_table.partitions(include_sub_partitions: true, include_self: true)
+      nested_target_mode = nested_target_tables.first.mode
 
-    successfully_acquired_lock = false
+      if nested_target_mode < target_mode
+        raise PgHaMigrations::InvalidMigrationError,
+          "Lock escalation detected! Cannot change lock level from :#{nested_target_mode} " \
+          "to :#{target_mode} for #{target_tables.map(&:fully_qualified_name).join(", ")}."
+      end
+    end
 
     until successfully_acquired_lock
-      while (
+      loop do
+        # If in a nested context and all of the above checks have passed,
+        # we have already acquired the lock so this check is unnecessary.
+        # In fact, it could actually cause a deadlock if a blocking query
+        # was executed shortly after the initial lock acquisition.
+        break if nested_target_tables
+
         blocking_transactions = PgHaMigrations::BlockingDatabaseTransactions.find_blocking_transactions("#{PgHaMigrations::LOCK_TIMEOUT_SECONDS} seconds")
-        blocking_transactions.any? do |query|
+
+        # Locking a partitioned table will also lock child tables (including sub-partitions),
+        # so we need to check for blocking queries on those tables as well
+        target_tables_for_blocking_transactions = target_tables.flat_map do |target_table|
+          target_table.partitions(include_sub_partitions: true, include_self: true)
+        end
```

Comment:
I moved this here so that we get a fresh list of tables for every iteration of the loop. I think this gives us extra safety if new partitions are added while this method is executing.